components/browser/domains/src/main/java/mozilla/components/browser/domains/Providers.kt

```diff
@@ -0,0 +1,114 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.domains
+
+import android.content.Context
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.async
+import kotlinx.coroutines.launch
+import mozilla.components.concept.toolbar.Autocomplete
+import java.util.Locale
+
+/**
+ * Provides autocomplete functionality for domains based on provided list of assets (see [Domains]).
+ */
+class ShippedDomainsProvider : Provider() {
+    override fun initialize(context: Context) {
+        scope.launch {
+            domains = async { Domains.load(context).into() }.await()
+        }
+    }
+}
+
+/**
+ * Provides autocomplete functionality for domains based on a list managed by [CustomDomains].
+ */
+class CustomDomainsProvider : Provider() {
+    override fun initialize(context: Context) {
+        scope.launch {
+            domains = async { CustomDomains.load(context).into() }.await()
+        }
+    }
+}
+
+/**
+ * Provides common autocomplete functionality powered by domain lists.
+ */
+abstract class Provider : Autocomplete {
```

Comment:
I'd find `AutocompleteProvider` and `DomainAutocompleteProvider` (for the common subtype) easier to read.